lib/clientv2/ippool.go

```diff
@@ -88,6 +96,61 @@ func (r ipPools) Update(ctx context.Context, res *apiv2.IPPool, opts options.Set
 
 // Delete takes name of the IPPool and deletes it. Returns an error if one occurs.
 func (r ipPools) Delete(ctx context.Context, name string, opts options.DeleteOptions) (*apiv2.IPPool, error) {
+	// Deleting a pool requires a little care because of existing endpoints
+	// using IP addresses allocated in the pool.  We do the deletion in
+	// the following steps:
+	// -  disable the pool so no more IPs are assigned from it
+	// -  remove all affinities associated with the pool
+	// -  delete the pool
+
+	// Get the pool so that we can find the CIDR associated with it.
+	pool, err := r.Get(ctx, name, options.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	logCxt := log.WithFields(log.Fields{
+		"CIDR": pool.Spec.CIDR,
+		"Name": name,
+	})
+
+	// If the pool is active, set the disabled flag to ensure we stop allocating from this pool.
```

Comment:
Why disable the pool if we are just about to delete it anyway?  Is this needed to allow ReleasePoolAffinities to work?

Reply:
hmmm, tries to remember the dim and distant past.. (this bit was copied over from the old client)

I believe this is to reduce the window of new affine blocks being claimed from the pool during the affine block cleanup which involves multiple read/writes to etcd.

However, there is still a window where it can happen.  Short of introducing more locking and more reads/writes during the IPAM affine block claiming code I think it'll be super hard to close this window completely - but this certainly reduces chances.  Also, claiming affine blocks should be a lot rarer than claiming IPs - and I don't believe we had any known issues with this approach since it was added to the last client.  But yes, perhaps we don't gain a whole lot from Disabling first.